### ISSUE TITLE
Need graph-safe multi-input TransformerBlock support for diffusion-style models with PP and CUDA graphs

### ISSUE THREAD

**Issue by whaletech-cheng** [NVIDIA/Megatron-LM#5620]
**Is your feature request related to a problem? Please describe.**

Diffusion-style transformer models often require auxiliary conditioning tensors, e.g. timestep embeddings for AdaLN/DiT blocks, text/image conditioning, or other per-layer inputs. These tensors need to be passed through every transformer layer and, with pipeline parallelism, may also need to cross PP stage boundaries with gradients.

Tag @NVIDIA/mcore-oncall.

Today, custom models can subclass `TransformerLayer` and override `get_layer_static_inputs()`, but end-to-end support is still difficult:

- `TransformerBlock.forward()` has a fixed layer-call signature and does not provide a generic way to pass custom Tensor kwargs such as `condition_emb` to all layers.
- TE CUDA graph capture requires all graph inputs to be Tensor/None and needs matching static sample kwargs.
- Pipeline schedules and tensor-shape helpers are primarily centered around a single hidden-state activation tensor. Multi-input or named activation transport requires custom hooks or packing workarounds.
- Recomputing conditioning modules independently on each PP stage can change model semantics for DiT-style models where a shared conditioning module should feed all layers.

This affects diffusion language models, DiT/AdaLN models, and multimodal models that are not simple GPT-style single-hidden-stream models.

**Describe the solution you'd like**

It would be useful to have an official graph-safe multi-input extension path for transformer blocks, for example:

1. Allow `TransformerBlock.forward()` to pass a generic `extra_layer_kwargs: dict[str, Tensor]` or named Tensor kwargs through to every layer.
2. Extend `get_layer_static_inputs()` / `TECudaGraphHelper` to support custom static Tensor kwargs in a documented way.
3. Provide a graph-safe packing/unpacking convention for non-Tensor metadata and Tensor-only CUDA graph inputs.
4. Provide PP support for named/list activation tensors, including tensor shape metadata and backward gradient propagation.
5. Add a small example or test for a DiT/AdaLN-style block with inputs like:
   - `hidden_states`
   - `attention_mask`
   - `packed_seq_params` / THD metadata
   - `condition_emb`

**Describe alternatives you've considered**

Current workarounds include:
- Copying `TransformerBlock.forward()` and manually adding custom kwargs.
- Packing auxiliary tensors into `hidden_states` for PP transport.
- Recomputing conditioning modules on every PP stage.
- Disabling PP or CUDA graphs for diffusion-style models.

These workarounds are brittle and make it harder to keep custom models compatible with future Megatron-Core / Megatron-Bridge versions.

**Additional context**

Megatron-Bridge diffusion examples provide useful local patterns, but they do not appear to demonstrate an end-to-end TP + PP + CP + TE CUDA graph path for a multi-input DiT-style transformer.

**Comment by Connor-XY** [comment]
Thanks for laying out the diffusion/DiT use case. The gap is real, but the request spans three separate interfaces: passing custom Tensor inputs through `TransformerBlock` and its layers, representing those inputs in TE CUDA-graph capture/cache keys, and transporting named auxiliary activations and gradients across pipeline stages.

This also overlaps the existing diffusion/video-model work tracked in #1592 and #2796. Routing this to @NVIDIA/pipeline-parallelism, @NVIDIA/cuda-graphs, and @NVIDIA/multi-modal.

Rather than addressing all three layers in one PR, the first reviewable slice should define a backward-compatible block/layer extension point for extra Tensor kwargs, with focused unit coverage. CUDA-graph integration and PP transport can then build on that agreed contract. Please outline the proposed API and static-versus-dynamic input semantics before starting the broader implementation.

**Comment by whaletech-cheng** [comment]
Thanks for the clarification. I agree that the first useful step should be a narrow, backward-compatible block/layer extension point for extra Tensor inputs, rather than trying to solve CUDA graph capture and PP transport in the same change.

For the first slice, the API I had in mind is:

```python
TransformerBlock.forward(
    hidden_states,
    attention_mask,
    ...,
    extra_layer_kwargs: Optional[dict[str, torch.Tensor | None]] = None,
)
```

`TransformerBlock` would pass these kwargs through to each layer:

```python
hidden_states, context = layer(
    hidden_states=hidden_states,
    attention_mask=attention_mask,
    context=context,
    ...,
    **(extra_layer_kwargs or {}),
)
```

For a DiT/AdaLN-style model, this would allow a conditioning tensor such as:

```python
extra_layer_kwargs = {
    "condition_emb": condition_emb,
}
```

where `condition_emb` is consumed by each custom transformer layer for modulation.

The intended semantics are:

- `extra_layer_kwargs=None` preserves the current behavior.
- Values are restricted to `torch.Tensor` or `None`.
- Non-Tensor metadata should remain in config/module attributes.
- The key set should be stable for a given model/forward path.
- Tensor values may change per microbatch.
- For future CUDA graph support, key set, Tensor-vs-None, shape, dtype, device, and requires_grad would be part of the static graph signature.
- Activation recompute should pass these kwargs through checkpointed layer calls so gradients to auxiliary Tensor inputs are preserved.
- PP transport semantics are out of scope for this first slice.

This would cover the minimal DiT use case without changing existing GPT-style model behavior, and it gives CUDA graph and PP support a clearer contract to build on later.

**Comment by Connor-XY** [comment]
Thanks, this proposed first slice looks like the right level of scope to me.

The key boundaries I would keep for a first PR are:

- restrict `extra_layer_kwargs` values to `torch.Tensor` or `None`, as you proposed;
- keep the key set stable for a given model forward path;
- pass the kwargs through activation recompute/checkpointed layer calls so gradients to auxiliary tensors are preserved;
- leave CUDA graph signature/cache-key handling and PP transport out of the initial PR.

Please open the first PR as a narrow backward-compatible `TransformerBlock` / layer extension point with focused unit coverage. In the PR description, call out the static-vs-dynamic contract you listed here, especially key-set stability and Tensor-vs-None semantics, so CUDA graph and PP follow-ups have a clear contract to build on.

**Comment by cspades** [comment]
Thanks Claude, clearly I'm an expert in diffusion models and PP just cause maybe I've said one too many words about CUDA graphs in public... /s
